contributors/design-proposals/local-pv-provisioner.md

```diff
@@ -0,0 +1,92 @@
+
+The master keeps track of `MinionInfo`, a collection containing one entry for each minion keeping track of relevant information. The master also runs a group of goroutines, one for each minion, which controls and interacts with the minions.
+
+### Minion Provisioner
+The minion provisioner exposes the following RPCs:
+
+| RPC                           | Description |
+| --- | --- |
+| `GetMinionName()`             | Returns the minion provisioner's name. |
+| `GetMinionStatus()`           | Returns the minion status, as described in the next section. |
+| `Delete(PersistentVolume)`    | Starts the asynchronous delete of the given PV. |
+
+In addition, the Discoverer and Deleter from the previous design are kept here, but they only perform filesystem operations. All other logic is moved to the master.
+
+### MinionStatus
+The following fields are retrieved as part of the `GetMinionStatus()` call:
+
+| Field               | Description |
+| --- | --- |
+| `LocalPVs`          | A collection of PVs mounted in the local storage directory on the node. The discoverer is triggered to acquire this information. |
+| `DeletesInProgress` | A collection of PVs currently being deleted asynchronously. This status is managed by the cleaner. |
+
+Every minion must maintain the invariant that `DeletesInProgress` is correct even after a minion restart. In the case that a user manually deletes the PV currently being deleted by a minion, master can only depend on the MinionStatus to ensure a replacement PV isn't created.
+
+### MinionInfo
+Each MinionInfo entry contains the following fields:
+
+| RPC                   | Description |
+| --- | --- |
+| `Name`                | Name of the minion provisioner. |
+| `Endpoint`            | The minion's host address. |
+| `APICache`            | Cache of API PV objects managed by this minion. |
+| `StopCh`              | Used to terminate a goroutine. |
+
+### Goroutine
+The goroutine will start by populating the `apiCache` using the populator logic that already exists in the previous design. Then, it will run the following loop with several seconds of delay in between executions:
+1. Call the RPC `GetMinionStatus()`.
+1. Using `LocalPVs` from the minion status, **reconcile** with the `APICache`. Create an additional PV if it exists in `LocalPVs` but not in `APICache` *and no delete is in progress for this PV, i.e. it's not in `DeletesInProgress`*.
+1. For each PV in `VolumeReleased` state (ready for cleanup), if it's not in `DeletesInProgress`, call the `Delete()` RPC. **TODO**:  How to delete PV object after disk delete operation is complete? Depends on cleaner implementation.
+
+**TODO**: gRPC backoff strategy in the case of network partition or minion failure.
+
+### Minion Discovery
+The master needs to have the host address of every minion in order to communicate with them directly. To do this, minions expose a single Service, which is associated with an Endpoints object containing all minion addresses.
```

Comment:
how is this different from the node API object, which contains addresses in status?

Reply:
The minions run as Pods, so will have IPs different from the Node